### PR TITLE
Added per round event-based setup in rapid tests

### PR DESCRIPTION
**Pull request by begmaroman**
# Description

Generate random setup per rounds in rapid testing 

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have added sufficient documentation in code

### CODE REVIEW

File: core/rapid_test.go [hunk 1]
```diff
@@ -51,48 +53,157 @@ func (m *mockInsertedProposals) insertProposal(
...
+		currentHeight: map[int]uint64{},
+		currentRound:  map[int]uint64{},
+	}
+
+	for height := uint64(0); height < desiredHeight; height++ {
+		var round uint64
+		for {
+			numByzantineNodes := rapid.
+				Uint64Range(0, maxBadNodes).
+				Draw(t, fmt.Sprintf("number of byzantine nodes for height %d on round %d", height, round))
```
Comment (JekaMas) [inline core/rapid_test.go:174]: I am not sure that it'll result in reproducible rapid test. You might need later to switch to generating slices.

**Comment by begmaroman** [comment]
This approach behaves more realistic than the previous one which is good. It generates random behavior per round until reaching consensus, then switching to the next height.

Once this PR is merged, I'll apply some more changes that would make the event generation logic more random. For instance, it would generate different messages for different nodes and see how it behaves, etc.

File: core/rapid_test.go (hunk 1 above)
Comment (begmaroman) [reply core/rapid_test.go:179]: It generates random data every time so everything should be ok

**Review by JekaMas — APPROVED** [review]

**Review by vcastellm — APPROVED** [review]
LGTM

**Pull request merged**